crates/bevy_picking_core/src/focus.rs

```diff
@@ -0,0 +1,141 @@
+//! Determines which entities are being hovered by pointers, taking into account [`FocusPolicy`],
+//! [`PickLayer`], and entity depth.
+
+use std::collections::BTreeMap;
+
+use crate::{backend, output::PointerInteraction, pointer::PointerId};
+use bevy::{
+    prelude::*,
+    ui::FocusPolicy,
+    utils::{FloatOrd, HashMap, HashSet},
+};
+
+/// A map of entities sorted by depth.
+type DepthMap = BTreeMap<FloatOrd, Entity>;
+
+/// Maps [`PickLayer`]s to the map of entities within that pick layer, sorted by depth.
+type LayerMap = BTreeMap<PickLayer, DepthMap>;
+
+/// Assigns an entity to a picking layer. When computing picking focus, entities
+/// are sorted in order from the highest to lowest layer, and by depth within each layer.
+#[derive(Debug, Clone, Copy, Component, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub enum PickLayer {
+    /// Topmost picking layer
+    Top = 0,
+    /// Layer immediately above the UI layer
+    AboveUi = 1,
+    /// Contains all UI entities
+    UI = 2,
+    /// Layer immediately below the UI
+    BelowUi = 3,
+    /// Layer immediately above the world
+    AboveWorld = 4,
+    /// The default `PickLayer`.
+    #[default]
+    World = 5,
+    /// Layer immediately below the world
+    BelowWorld = 6,
+    /// Bottommost picking layer
+    Bottom = 7,
+}
+
+/// Maps Pointers to a [`LayerMap`]. Note this is much more complex than the [`HoverMap`] because
+/// this data structure is used to sort entities by layer then depth for every pointer.
+type OverMap = HashMap<PointerId, LayerMap>;
+
+/// Maps pointers to the entities they are hovering over. "Hovering" refers to the actual hover
+/// interaction state, in this case, not just whether the pointer happens to be over the entity.
+#[derive(Debug, Deref, DerefMut, Default)]
+pub struct HoverMap(pub HashMap<PointerId, HashSet<Entity>>);
+
+/// Coalesces all data from inputs and backends to generate a map of the currently hovered entities.
+pub fn update_focus(
+    // Inputs
+    focus: Query<&FocusPolicy>,
+    pick_layers: Query<&PickLayer>,
+    pointers: Query<(&PointerId, &PointerInteraction)>, // <- what happened last frame
```

Comment:
This function would benefit from more extensive docs.

Reply:
I'm not quite sure what to add here, the function is only three lines long, and each of those function document what they do. I'd like to avoid duplicating information. I added another line of context to the comment.